### PR TITLE
[LDR-R1] EOS:13790:S3: Support bundle optimization

### PR DESCRIPTION
**Pull request by sachitanands**
Support bundle optimization,
1) set limit to core files as 11
2) Instead of m0trace files we can collect its output m0trace -i m0trace.118914 > out.txt
     (Instead of txt we can use yaml file suggested by motr team)
3) Get netstat for port numbers : netstat -tulpn | grep -i listen
4) Get rpm package names and versions : rpm -qa | grep cortx

Test plan,
1) Test support bundle on empty directory 
2) Test support on h/w node multiple time : record time and support bundle size (run it 2-3 times)
3) validate you can do unxz and untar operation
    e.g. 
	 1) unxz <sb>.tar.gz
	 2) tar -xf <step 1 output> -C <dest-path>

4) Test support bundle with old/invalid s3 binary (gdb behaviour test)
5) (if possible) test m0tracedump command exceptions wiht support bundle

Signed-off-by: Sachitanand Shelake <sachitanand.shelake@seagate.com>

### CODE REVIEW

**Review by abhilekh268 — APPROVED** [review]

File: scripts/s3-support-bundles/s3_bundle_generate.sh [hunk 1]
```diff
@@ -98,67 +100,105 @@ s3server_logdir=`cat $s3server_config | grep "S3_LOG_DIR:" | cut -f2 -d: | sed -
...
-  done
+# Collect call stack of latest <s3_core_files_max_count> s3server core files
+# from s3_core_dir directory, if available
+collect_core_files(){
+  core_filename_pattern="core-s3server.*.gz"
+  mkdir -p $s3_core_files
+  cwd=$(pwd)
+  cd $s3_core_dir
+  # get recent modified core files from directory
+  (ls -t $core_filename_pattern 2>/dev/null | head -$s3_core_files_max_count) | xargs -I '{}' cp '{}' $s3_core_files
```
Comment (nileshgovande) [inline scripts/s3-support-bundles/s3_bundle_generate.sh:111]: use \ls 
generally use unaliased version

File: scripts/s3-support-bundles/s3_bundle_generate.sh [hunk 2]
```diff
@@ -98,67 +100,105 @@ s3server_logdir=`cat $s3server_config | grep "S3_LOG_DIR:" | cut -f2 -d: | sed -
...
+  else
+      # iterate over the s3_core_files directory and extract call stack for each file
+      for s3corefile in "$s3_core_files"/*
+      do
+          zipped_core_name=$(basename "$s3corefile")
+          core_name=${zipped_core_name%".gz"}
+          callstack_file="$s3_core_files"/callstack."$core_name".txt
+
+          # unzip the core file, to read it using gdb
+          tar -xzf $s3corefile -C "$s3_core_files"
```
Comment (nileshgovande) [inline scripts/s3-support-bundles/s3_bundle_generate.sh:126]: use gunzip

File: scripts/s3-support-bundles/s3_bundle_generate.sh (hunk 1 above)
Comment (nileshgovande) [reply scripts/s3-support-bundles/s3_bundle_generate.sh:111]: No need to copy

File: scripts/s3-support-bundles/s3_bundle_generate.sh [hunk 3]
```diff
@@ -98,67 +100,105 @@ s3server_logdir=`cat $s3server_config | grep "S3_LOG_DIR:" | cut -f2 -d: | sed -
...
+
+          printf "Core name: $core_name\n" >> "$callstack_file"
+          printf "Callstack:\n\n" >> "$callstack_file"
+
+          # generate gdb bt and append into the callstack_file
+          gdb --batch --quiet -ex "thread apply all bt full" -ex "quit" $s3server_binary\
```
Comment (nileshgovande) [inline scripts/s3-support-bundles/s3_bundle_generate.sh:132]: For big core files, use gzip -l, figure out the 'inflated' size and only inflate if < 10GB , else just make a placeholder file

**Comment by nileshgovande** [comment]
use xz in the supportbundle filename

File: scripts/s3-support-bundles/s3_bundle_generate.sh (hunk 3 above)
Comment (sachitanands) [reply scripts/s3-support-bundles/s3_bundle_generate.sh]: done

File: scripts/s3-support-bundles/s3_bundle_generate.sh (hunk 2 above)
Comment (sachitanands) [reply scripts/s3-support-bundles/s3_bundle_generate.sh]: done

**Review by abhilekh268 — APPROVED** [review]

**Comment by nileshgovande** [comment]
No need of running UTs and STs since support bundle change only

**Review by amitwac2608 — APPROVED** [review]
looks good.

**Pull request merged**